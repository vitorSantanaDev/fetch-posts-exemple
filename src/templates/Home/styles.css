.searchContainer {
  margin-bottom: 40px;
}
.searchContainer h1 {
  margin-bottom: 32px;
}