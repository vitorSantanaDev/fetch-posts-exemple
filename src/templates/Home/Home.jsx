import { useCallback, useEffect, useState } from 'react';
import './styles.css';
import { loadPosts } from '../../utils/loadPosts.js';
import Container from '../../components/Container';
import Posts from '../../components/Posts';
import Button from '../../components/Button';
import Input from '../../components/Input';

const Home = () => {
  const [page, setPage] = useState(0);
  const [postsPerPage] = useState(3);
  const [allPosts, setAllPosts] = useState([]);
  const [posts, setPosts] = useState([]);
  const [searchValue, setSearchValue] = useState('');

  const loadMorePosts = () => {
    const nextPage = page + postsPerPage;
    const nextPosts = allPosts.slice(nextPage, nextPage + postsPerPage);

    posts.push(...nextPosts);
    setPosts(posts);
    setPage(nextPage);
  };

  const loadPostsInTheHome = useCallback(async (page, postsPerPage) => {
    const postsAndPhotos = await loadPosts();
    setPosts(postsAndPhotos.slice(page, postsPerPage));
    setAllPosts(postsAndPhotos);
  }, []);

  const handleChangeInputField = ({ target }) => {
    setSearchValue(target.value);
  };

  useEffect(() => {
    loadPostsInTheHome(0, postsPerPage);
  }, [loadPostsInTheHome, postsPerPage]);

  const noMorePosts = page + postsPerPage >= allPosts.length;
  const filteredPosts = searchValue
    ? allPosts.filter((post) => {
        return post.title.toLowerCase().includes(searchValue.toLowerCase());
      })
    : posts;

  return (
    <Container>
      <div className="searchContainer">
        {!!searchValue && <h1>Search: {searchValue}</h1>}
        <Input value={searchValue} onChange={handleChangeInputField} />
      </div>

      {filteredPosts.length > 0 && <Posts posts={filteredPosts} />}
      {filteredPosts.length === 0 && <h1>Post n√£o encontrado! :(</h1>}
      {!searchValue && <Button disabled={noMorePosts} onClick={loadMorePosts} text="Load More Posts" />}
    </Container>
  );
};

export default Home;
