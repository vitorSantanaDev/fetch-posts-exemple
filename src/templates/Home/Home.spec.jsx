import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import Home from './Home';

const handlers = [
  rest.get('*jsonplaceholder.typicode*', async (req, res, ctx) => {
    return res(
      ctx.json([
        {
          userID: 1,
          id: 1,
          title: 'Title1 one!',
          body: 'Body post one, more reading posts.',
          url: 'img1.jpg',
        },
        {
          userID: 2,
          id: 2,
          title: 'Title2 two!',
          body: 'Body post two, more reading posts.',
          url: 'img2.jpg',
        },
        {
          userID: 3,
          id: 3,
          title: 'Title3 three!',
          body: 'Body post three, more reading posts.',
          url: 'img3.jpg',
        },
        {
          userID: 4,
          id: 4,
          title: 'Title4 for!',
          body: 'Body post for, more reading posts.',
          url: 'img4.jpg',
        },
        {
          userID: 5,
          id: 5,
          title: 'Title5 five!',
          body: 'Body post for, more reading posts.',
          url: 'img4.jpg',
        },
      ]),
    );
  }),
];
const server = setupServer(...handlers);

describe('<Home />', () => {
  beforeAll(() => {
    server.listen();
  });

  afterEach(() => server.resetHandlers());

  afterAll(() => {
    server.close();
  });

  it('should render search Input, Posts and loadMorePosts', async () => {
    render(<Home />);
    const noMorePosts = screen.getByText('Post não encontrado! :(');
    await waitForElementToBeRemoved(noMorePosts);

    expect.assertions(3);

    const search = screen.getByPlaceholderText(/type your search/i);
    expect(search).toBeInTheDocument();

    const images = screen.getAllByRole('img', { name: /title/i });
    expect(images).toHaveLength(4);

    const button = screen.getByRole('button', { name: /load more posts/i });
    expect(button).toBeInTheDocument();
  });

  it('should search for posts', async () => {
    render(<Home />);
    const noMorePosts = screen.getByText('Post não encontrado! :(');
    await waitForElementToBeRemoved(noMorePosts);

    expect.assertions(13);

    const search = screen.getByPlaceholderText(/type your search/i);

    expect(screen.getByRole('heading', { name: 'Title1 one!' })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: 'Title2 two!' })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: 'Title3 three!' })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: 'Title4 for!' })).toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: 'Title5 five!' })).not.toBeInTheDocument();

    userEvent.type(search, 'title1');
    expect(screen.getByRole('heading', { name: 'Title1 one!' })).toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: 'Title2 two!' })).not.toBeInTheDocument();
    expect(screen.queryByRole('heading', { name: 'Title3 three!' })).not.toBeInTheDocument();

    expect(screen.getByRole('heading', { name: 'Search: title1' })).toBeInTheDocument();

    userEvent.clear(search);
    expect(screen.getByRole('heading', { name: 'Title1 one!' })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: 'Title2 two!' })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: 'Title3 three!' })).toBeInTheDocument();

    userEvent.type(search, 'post do not exist');
    expect(screen.getByText('Post não encontrado! :(')).toBeInTheDocument();
  });

  it('should load more posts', async () => {
    render(<Home />);
    const noMorePosts = screen.getByText('Post não encontrado! :(');

    expect.assertions(2);
    await waitForElementToBeRemoved(noMorePosts);
    const button = screen.getByRole('button', { name: /load more posts/i });

    userEvent.click(button);

    expect(screen.queryByRole('heading', { name: 'Title5 five!' })).toBeInTheDocument();
    expect(button).toBeDisabled();
  });

  it('should match snapshopt', async () => {
    const { container } = render(<Home />);
    const noMorePosts = screen.getByText('Post não encontrado! :(');
    await waitForElementToBeRemoved(noMorePosts);

    expect(container.firstChild).toMatchSnapshot();
  });
});
