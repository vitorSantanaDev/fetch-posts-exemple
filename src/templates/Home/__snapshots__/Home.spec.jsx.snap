// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should match snapshopt 1`] = `
<div
  class="container"
>
  <div
    class="searchContainer"
  >
    <input
      class="textInput"
      placeholder="Type your search"
      type="search"
      value=""
    />
  </div>
  <div
    class="posts"
  >
    <div
      class="post"
    >
      <img
        alt="Title1 one!"
        src="img1.jpg"
      />
      <div
        class="postContent"
      >
        <h2>
          Title1 one!
        </h2>
        <p>
          Body post one, more reading posts.
        </p>
      </div>
    </div>
    <div
      class="post"
    >
      <img
        alt="Title2 two!"
        src="img2.jpg"
      />
      <div
        class="postContent"
      >
        <h2>
          Title2 two!
        </h2>
        <p>
          Body post two, more reading posts.
        </p>
      </div>
    </div>
    <div
      class="post"
    >
      <img
        alt="Title3 three!"
        src="img3.jpg"
      />
      <div
        class="postContent"
      >
        <h2>
          Title3 three!
        </h2>
        <p>
          Body post three, more reading posts.
        </p>
      </div>
    </div>
    <div
      class="post"
    >
      <img
        alt="Title4 for!"
        src="img4.jpg"
      />
      <div
        class="postContent"
      >
        <h2>
          Title4 for!
        </h2>
        <p>
          Body post for, more reading posts.
        </p>
      </div>
    </div>
  </div>
  <div
    class="buttonContainer"
  >
    <button
      class="button"
    >
      Load More Posts
    </button>
  </div>
</div>
`;
